### ISSUE TITLE
Rules pertaining to PEP 695

### ISSUE THREAD

**Issue by Goldziher** [astral-sh/ruff#12542]
MyPy released support for PEP 695 generics: https://github.com/python/mypy/issues/15238

It would be awesome to do any of the following:

1. get linting errors for code that uses older style generics (TypeVar, ParamSpec etc.) for Python 3.12+ 
2. get linting errors for code that uses newer style generics and should be backward compatible with lower Python versions
3. automagically upgrade older style generics to use the new syntax 

I see that for point 3 there is an [open related issue](https://github.com/astral-sh/ruff/issues/4617). This issue is about the underlying rust implementation for point 3, but does not cover points 1 and 2.

**Comment by charliermarsh** [comment]
We have some support for this already in https://docs.astral.sh/ruff/rules/non-pep695-type-alias/. I believe that rule does (1) and (3)?

**Comment by Goldziher** [comment]
> We have some support for this already in https://docs.astral.sh/ruff/rules/non-pep695-type-alias/. I believe that rule does (1) and (3)?

I don't think so. I might be wrong but it seems to me like this rule is concerned with the `TypeAlias` type. 

I am talking about the new generics style: https://peps.python.org/pep-0695/#summary-examples

**Comment by charliermarsh** [comment]
Thanks, you're right, that's my mistake. We may want to fold this into #4617, and just flesh out that issue a bit more.
